catalog/IndexScheme.cpp

```diff
@@ -0,0 +1,101 @@
+/**
+ *   Copyright 2016, Quickstep Research Group, Computer Sciences Department,
+ *     University of Wisconsin—Madison.
+ *
+ *   Licensed under the Apache License, Version 2.0 (the "License");
+ *   you may not use this file except in compliance with the License.
+ *   You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *   Unless required by applicable law or agreed to in writing, software
+ *   distributed under the License is distributed on an "AS IS" BASIS,
+ *   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *   See the License for the specific language governing permissions and
+ *   limitations under the License.
+ **/
+
+#include "catalog/IndexScheme.hpp"
+
+#include <memory>
+#include <string>
+#include <unordered_map>
+#include <utility>
+#include <vector>
+
+#include "catalog/Catalog.pb.h"
+#include "catalog/CatalogTypedefs.hpp"
+#include "storage/StorageBlockInfo.hpp"
+#include "threading/SpinSharedMutex.hpp"
+#include "types/TypedValue.hpp"
+#include "types/operations/comparisons/Comparison.hpp"
+#include "types/operations/comparisons/LessComparison.hpp"
+#include "utility/Macros.hpp"
+
+#include "glog/logging.h"
+
+namespace quickstep {
+
+bool IndexScheme::ProtoIsValid(const serialization::IndexScheme &proto) {
+  // Check that proto is fully initialized.
+  if (!proto.IsInitialized()) {
+    return false;
+  }
+
+  // Check that each index entry contains initialized valid index descriptions.
+  for (int i = 0; i < proto.index_entries_size(); ++i) {
+    const serialization::IndexScheme_IndexEntry &index_entry = proto.index_entries(i);
+    for (int j = 0; j < index_entry.index_description_size(); ++j) {
+      const IndexSubBlockDescription &index_description = index_entry.index_description(j);
+      if (!index_description.IsInitialized()) {
```

Comment:
I believe I have mentioned before that such check for `IndexSubBlockDescription` is not sufficient enough for three implementations. Instead, we need to add a new static method called `ProtoIsValid` (maybe in a new file `/storage/IndexSubBlockFactory` to avoid cycle dependency).

Have you added that? Thanks!